### PR TITLE
Pacbio variantcalling

### PR DESCRIPTION
**Pull request by Redmar-van-den-Berg**
### Checklist
- [x] Pull request details were added to CHANGELOG.md
- [x] `parameter_meta` for each task is up to date.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,11 @@ that users understand how the changes affect the new version.
 
 version 5.0.0-dev
 ---------------------------
++ deepvariant: Add task for DeepVariant
```
Comment (JasperBoom) [inline CHANGELOG.md:14]:
```suggestion
+ deepvariant: Add task for DeepVariant.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,11 @@ that users understand how the changes affect the new version.
...
++ gatk: Make intervals optional for GenotypeGVCFs
```
Comment (JasperBoom) [inline CHANGELOG.md:15]:
```suggestion
+ gatk: Make intervals optional for GenotypeGVCFs.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,6 +11,11 @@ that users understand how the changes affect the new version.
...
++ isoseq3: Add required bam index input to isoseq3
```
Comment (JasperBoom) [inline CHANGELOG.md:16]:
```suggestion
+ isoseq3: Add required bam index input to isoseq3.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -11,6 +11,11 @@ that users understand how the changes affect the new version.
...
++ pbbam: Add task for indexing PacBio bam files
```
Comment (JasperBoom) [inline CHANGELOG.md:17]:
```suggestion
+ pbbam: Add task for indexing PacBio bam files.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -11,6 +11,11 @@ that users understand how the changes affect the new version.
...
++ picard: Add CollectHsMetrics and CollectVariantCallingMetrics
```
Comment (JasperBoom) [inline CHANGELOG.md:18]:
```suggestion
+ picard: Add CollectHsMetrics and CollectVariantCallingMetrics.
```

File: isoseq3.wdl [hunk 6]
```diff
@@ -26,6 +26,7 @@ task Refine {
         Boolean requirePolyA = false
         String logLevel = "WARN"
         File inputBamFile
+        File inputBamIndex
```
Comment (JasperBoom) [inline isoseq3.wdl:29]: Also add a parameter_meta for this new input.

File: picard.wdl [hunk 7]
```diff
@@ -66,6 +66,70 @@ task BedToIntervalList {
...
+        File inputBam
+        File inputBamIndex
+        File referenceFasta
+        File referenceFastaDict
+        File referenceFastaFai
+        File targets
+        File? baits
+        String basename
+
+
```
Comment (JasperBoom) [inline picard.wdl:80]:
```suggestion
```

File: picard.wdl [hunk 8]
```diff
@@ -66,6 +66,70 @@ task BedToIntervalList {
...
 }
 
+task CollectHsMetrics {
+    input {
+        File inputBam
+        File inputBamIndex
+        File referenceFasta
+        File referenceFastaDict
+        File referenceFastaFai
+        File targets
```
Comment (JasperBoom) [inline picard.wdl:76]: Missing parameter_meta here.

File: picard.wdl [hunk 9]
```diff
@@ -66,6 +66,70 @@ task BedToIntervalList {
...
+        File? baits
```
Comment (JasperBoom) [inline picard.wdl:77]: Missing parameter_meta here.

File: picard.wdl [hunk 10]
```diff
@@ -315,6 +379,53 @@ task CollectTargetedPcrMetrics {
     }
 }
 
+task CollectVariantCallingMetrics {
+    input {
+        File dbsnp
```
Comment (JasperBoom) [inline picard.wdl:384]: Parameter_meta should include entry for all inputs.

File: lima.wdl [hunk 11]
```diff
@@ -58,7 +58,6 @@ task Lima {
 
     command {
         set -e
-        mkdir -p "$(dirname ~{outputPrefix})"
```
Comment (JasperBoom) [inline lima.wdl:61]: What is the reason for the changes here, going back to the glob structure?

Comment (Redmar-van-den-Berg) [reply lima.wdl:61]: It's a lot simpler, and since we no longer support multiple input bam files will be no name collisions.

File: whatshap.wdl [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+        File vcf
+
+        String memory = "4G"
+        Int timeMinutes = 120
+        # Whatshap 1.0, tabix 0.2.5
+        String dockerImage = "quay.io/biocontainers/mulled-v2-5c61fe1d8c284dd05d26238ce877aa323205bf82:89b4005d04552bdd268e8af323df83357e968d83-0"
+      }
+
+    command {
+      whatshap stats \
```
Comment (JasperBoom) [inline whatshap.wdl:171]:
```suggestion
        whatshap stats \
```

File: whatshap.wdl [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+        ~{if defined(tsv) then ("--tsv " +  '"' + tsv + '"') else ""} \
+        ~{true="--only-snvs" false="" only_sn_vs} \
+        ~{if defined(block_list) then ("--block-list " +  '"' + block_list + '"') else ""} \
+        ~{if defined(chromosome) then ("--chromosome " +  '"' + chromosome + '"') else ""}
+    }
+
+    output {
+      File? phasedGTF = gtf
+      File? phasedTSV = tsv
+      File? phasedBlockList = block_list
```
Comment (JasperBoom) [inline whatshap.wdl:185]:
```suggestion
        File? phasedGTF = gtf
        File? phasedTSV = tsv
        File? phasedBlockList = block_list
```

File: whatshap.wdl [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+task Phase {
+    input {
+        String outputVCF
+        File? reference
+        File? referenceIndex
+        Boolean? no_reference
```
Comment (JasperBoom) [inline whatshap.wdl:29]: https://biowdl.github.io/styleGuidelines.html#input-names

Inputs should be lowerCamelCase according to our BioWDL styleguide.

File: whatshap.wdl [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+        timeMinutes: {description: "The maximum amount of time the job will run in minutes.", category: "advanced"}
+        dockerImage: {description: "The docker image used for this task. Changing this may result in errors which the developers may choose not to address.", category: "advanced"}
+    }
+}
+
+task Stats {
+    input {
+        String? gtf
+        String? sample
+        String? chr_lengths
```
Comment (JasperBoom) [inline whatshap.wdl:157]: https://biowdl.github.io/styleGuidelines.html#input-names

Same here.

File: whatshap.wdl [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+    }
+}
+
+task Haplotag {
+    input {
+        String outputFile
+        File? reference
+        File? referenceFastaIndex
+        String? regions
+        Boolean? ignore_linked_read
```
Comment (JasperBoom) [inline whatshap.wdl:215]: https://biowdl.github.io/styleGuidelines.html#input-names

File: whatshap.wdl [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+          ~{if defined(outputFile) then ("--output " +  '"' + outputFile+ '"') else ""} \
+          ~{if defined(reference) then ("--reference " +  '"' + reference + '"') else ""} \
+          ~{if defined(regions) then ("--regions " +  '"' + regions + '"') else ""} \
+          ~{true="--ignore-linked-read" false="" ignore_linked_read} \
+          ~{if defined(linked_read_distance_cut_off) then ("--linked-read-distance-cutoff " +  '"' + linked_read_distance_cut_off + '"') else ""} \
+          ~{true="--ignore-read-groups" false="" ignore_read_groups} \
+          ~{if defined(sample) then ("--sample " +  '"' + sample + '"') else ""} \
+          ~{if defined(output_haplo_tag_list) then ("--output-haplotag-list " +  '"' + output_haplo_tag_list + '"') else ""} \
+          ~{true="--tag-supplementary" false="" tag_supplementary} && \
+          python3 -c "import pysam; pysam.index('~{outputFile}')"
```
Comment (JasperBoom) [inline whatshap.wdl:245]:
```suggestion
        whatshap haplotag \
        ~{vcf} \
        ~{alignments} \
        ~{if defined(outputFile) then ("--output " +  '"' + outputFile+ '"') else ""} \
        ~{if defined(reference) then ("--reference " +  '"' + reference + '"') else ""} \
        ~{if defined(regions) then ("--regions " +  '"' + regions + '"') else ""} \
        ~{true="--ignore-linked-read" false="" ignore_linked_read} \
        ~{if defined(linked_read_distance_cut_off) then ("--linked-read-distance-cutoff " +  '"' + linked_read_distance_cut_off + '"') else ""} \
        ~{true="--ignore-read-groups" false="" ignore_read_groups} \
        ~{if defined(sample) then ("--sample " +  '"' + sample + '"') else ""} \
        ~{if defined(output_haplo_tag_list) then ("--output-haplotag-list " +  '"' + output_haplo_tag_list + '"') else ""} \
        ~{true="--tag-supplementary" false="" tag_supplementary} && \
        python3 -c "import pysam; pysam.index('~{outputFile}')"
```

File: whatshap.wdl [hunk 18]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    output {
+      File bam = outputFile
+      File bamIndex = outputFile + ".bai"
```
Comment (JasperBoom) [inline whatshap.wdl:250]:
```suggestion
        File bam = outputFile
        File bamIndex = outputFile + ".bai"
```

File: whatshap.wdl [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    runtime {
+        docker: dockerImage
+        time_minutes: timeMinutes
+        memory: memory
+    }
+
+    parameter_meta {
```
Comment (JasperBoom) [inline whatshap.wdl:259]: Make sure all inputs are in the parameter_meta.

File: whatshap.wdl [hunk 20]
```diff
@@ -0,0 +1,275 @@
...
+        File phasedVCFIndex = outputVCF + ".tbi"
+    }
+
+    runtime {
+        docker: dockerImage
+        time_minutes: timeMinutes
+        memory: memory
+    }
+
+    parameter_meta {
```
Comment (JasperBoom) [inline whatshap.wdl:114]: Make sure all inputs are in the parameter_meta.

**Review by JasperBoom — CHANGES_REQUESTED** [review]

File: lima.wdl (hunk 11 above)
Comment (JasperBoom) [reply lima.wdl:61]: But you do get these glob folders in the output now right? Or does that not happen when its just one file per glob?

Also, shouldn't the task check the possible output directory if that is set in the main pipeline?
Since you've remove `mkdir -p "$(dirname ~{outputPrefix})"` now?

File: pbbam.wdl [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+
+task Index {
+    input {
+        File bamFile
+        String? outputBamPath
```
Comment (JasperBoom) [inline pbbam.wdl:25]:
```suggestion
        String? outputBamPath
        
```

Sorry about the nitpicking on spacing here, but we have such a line between the runtime inputs and the rest :see_no_evil:

File: lima.wdl (hunk 11 above)
Comment (Redmar-van-den-Berg) [reply lima.wdl:61]: I've also removed the outputPrefix from the subreads-processing pipeline, so that is not a problem. I've been using these tasks for a few weeks already, so they all work.

**Review by JasperBoom — APPROVED** [review]
